pkg/managed/clusterroles_test.go

```diff
@@ -0,0 +1,281 @@
+// Copyright (C) 2020, Oracle and/or its affiliates.
+// Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl.
+package managed
+
+import (
+	"context"
+	"github.com/stretchr/testify/assert"
+	"github.com/verrazzano/verrazzano-operator/pkg/constants"
+	"github.com/verrazzano/verrazzano-operator/pkg/monitoring"
+	"github.com/verrazzano/verrazzano-operator/pkg/testutil"
+	"github.com/verrazzano/verrazzano-operator/pkg/types"
+	"github.com/verrazzano/verrazzano-operator/pkg/util"
+	rbacv1 "k8s.io/api/rbac/v1"
+	v1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"reflect"
+	"testing"
+)
+
+// TestCleanupOrphanedClusterRoles tests the cleanup of orphaned cluster roles
+// GIVEN a cluster which has no existing cluster roles
+//  WHEN I create an orphaned cluster (not expected on the cluster) and call CleanupOrphanedClusterRoles
+//  THEN the orphaned cluster role should be cleaned up (deleted)
+func TestCleanupOrphanedClusterRoles(t *testing.T) {
+	modelBindingPair := testutil.GetModelBindingPair()
+	clusterConnections := testutil.GetManagedClusterConnections()
+	clusterConnection := clusterConnections["cluster3"]
+
+	// create a cluster role that is not expected on the cluster
+	cr := v1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "orphaned",
+			Labels: map[string]string{
+				"verrazzano.binding": "testBinding",
+				"verrazzano.cluster": "cluster3",
+			},
+		},
+	}
+	_, err := clusterConnection.KubeClient.RbacV1().ClusterRoles().Create(context.TODO(), &cr, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("got an error trying to create a cluster role: %v", err)
+	}
+
+	// the orphaned cluster role should be cleaned up by the call to CleanupOrphanedClusterRoles
+	err = CleanupOrphanedClusterRoles(modelBindingPair, clusterConnections)
+	if err != nil {
+		t.Fatalf("got an error from CleanupOrphanedClusterRoles: %v", err)
+	}
+
+	// verify that the orphaned cluster role no longer exists
+	_, err = clusterConnection.KubeClient.RbacV1().ClusterRoles().Get(context.TODO(), "orphaned", metav1.GetOptions{})
+	if err == nil {
+		t.Fatal("expected cluster role to be cleaned up")
+	}
+}
+
+// TestCreateClusterRoles tests creation of cluster roles
+// GIVEN a cluster which does not have a cluster role with the service account name
```

Comment:
this also made me stop an think - a clusterrole has nothing to do with a service account name, its a clusterrolebinding that associates a cluster role with a user, like a service account. the func being tested does not appear to touch SA's in any way - so really this is just any arbitrary name, it does not have to be a SA name, if i am reading this right. maybe we create clusterroles with the same name as a SA by convention, but i think this test description is implying that the clusterrole name must be the same as a SA name, which does not seem to be the case?

Reply:
The behavior of the function is that if there is no cluster role with the SA name then one is created with set of expected policy rules.  That's what I was trying to say with the description.  If that is incorrect or misleading I'm happy to change it.  Do you have any suggestions on how to improve it?  I don't pretend to know exactly why the code was written as it is.  I know it would be good to have a deeper understanding of the code but I'm trying to strike a balance between digging deeper and understanding enough to write unit tests that cover most paths and have a reasonable set of assertions.  Again, I'm not really sure what to do here.

Reply:
that's ok - like the other comment, you are just trying to write a test for the code that is already there.  let's just leave it as it.  that func does not seem to use an SA name, except maybe by accident.  it probably has more things wrong with it.  at least this test will create a guard for future changes...